Add buyable option to disable purchase confirmation

arinthros opened the pull request:
I got frustrated having to confirm every small purchase, so I created an upgrade to buy to let me disable the purchase confirmation screen. Putting it behind a 100k purchase makes it so new players don't accidentally buy massive things until they learn the game. 100k might be too high, not sure how upgrades are valued, but it felt when I could afford 100k is when I started to get annoyed with the confirmation.

This needs tests if you want to require it, but I confirmed it works to purchase the upgrade, and the store behaves correctly when enabled and disabled.

arinthros committed:
Add buyable option to disable purchase confirmation

rj-lee reviewed (commented):
A couple of thoughts:

- First of all, great job on the first attempt at legit python! The code looks good
- I pulled this down to test it, and I was expecting to be able to buy the next thing right after I bought something, but it brings me back to the [Buy, Sell, Go Back] screen so it only half-solves the problem of not being able to buy continuously. It might be nice to be able to just press enter enter enter to buy buy buy.
- When I first bought this upgrade I was hoping it would work for selling as well, then we could press enter enter enter to sell sell sell, but I know you only scoped it for purchases and not selling
- It feels a little weird that the upgrade we get is to "disable" something, I wonder if it would make sense to have the upgrade allow you to enable "purchase confirmation skipping", but this is just minor language difference
- I didn't think players really wanted to buy that many things from the normal store, is the main use case for this buying seeds and armor sets? I haven't had to buy a lot of other things that made it tedious

Mostly questions and thoughts about the functionality of the upgrade, but we'll see what @TheRedPanda17 thinks about whether this is a valuable and balanced addition to the game

TheRedPanda17 commented:
Yeah I think my main question is about the use-case. The code seems fine

arinthros commented:
The main use case is seeds, but garden is using the BaseShop class so I stuck the change in there, making it apply to all stores. Would it make sense to do it similar to the sorts, where it checks the shop type? Or has a level-able unlock? 25k store, 50k armory, 100k garden?

I agree it feels weird to go back to the previous menu and not stay in the shop. I'll add that tweak.

TheRedPanda17 commented:
@arinthros Have you unlocked the Does the Buy All Seeds option?

arinthros commented:
I haven't unlocked buy all seeds because I'm a snob and only buy premium :)

rj-lee commented:
ahhh so maybe what we actually need is an option to "buy highest value seeds" or "increase garden shop seed quality"?

TheRedPanda17 commented:
That makes more sense. I think maybe we should just add "confirm purchases" as an option from the start instead of an upgrade. Then people can toggle it if they want. Thoughts?

rj-lee commented:
There will be a disparity between having optional confirmation for buying and always having a confirmation for selling?

arinthros committed:
Simplify confirmation setting

- Remove upgrade option
- Flip logic & rename setting
- Stay on items list screen
- WIP: Select adjacent item instead of going to top of list

arinthros committed:
Select adjacent item when buying

arinthros commented:
> There will be a disparity between having optional confirmation for buying and always having a confirmation for selling?

Yes, but per one of our convos that might break the "sell all" upgrade mechanic by making it too easy to rapidly sell items.

arinthros commented on myning/chapters/base_store.py:
```diff
@@ -185,3 +205,15 @@ def is_useful_item(self, item: Item):
                 return True
         return False
+
+    def pick_preserve_cursor(self, args: PickArgs):
```
Big thanks to @rj-lee for helping figure this out.

arinthros committed:
Remove unused code

arinthros commented:
@rj-lee @TheRedPanda17 this should be ready to go now. Let me know if it needs any changes? I've been playing on this branch exclusively, and so far so good.

rj-lee commented:
> @rj-lee @TheRedPanda17 this should be ready to go now. Let me know if it needs any changes? I've been playing on this branch exclusively, and so far so good.

The code looks good to me so I'll just defer to @TheRedPanda17 for the go-ahead with the buy/sell dialogue disparity
